### PR TITLE
SALTO-1042: added go to service button in the extension

### PR DESCRIPTION
**Pull request by alonstern**
Implemented go to service right-click button in VSCode.

---
__Release Notes:__
VSCode Extension:
Added go to service right-click menu option to go to the service page in the browser of a certain element (Currently only supported for salesforce elements).

### CODE REVIEW

File: packages/lang-server/src/workspace.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ import { Workspace, nacl, errors, parser } from '@salto-io/workspace'
 import { Element, SaltoError, ElemID } from '@salto-io/adapter-api'
 
 export class EditorWorkspace {
-  private workspace: Workspace
+  readonly workspace: Workspace
```
Comment (ori-moisis) [inline packages/lang-server/src/workspace.ts:23]: should not allow direct access to the inner workspace.
IIRC the editor workspace has some kind of caching mechanism that means that allowing external access to the underlying workspace may cause consistency issues

@roironn @noamhammer - correct me if I am wrong about this

File: packages/vscode/package.json [hunk 2]
```diff
@@ -116,12 +125,15 @@
         "generate-yarn-lock": "yarn workspaces run generate-lock-entry | sed '1,2d' | sed -n -e :a -e '1,1!{P;N;D;};N;ba' >> yarn.lock"
     },
     "dependencies": {
+        "@salto-io/adapter-api": "0.2.0",
```
Comment (ori-moisis) [inline packages/vscode/package.json:128]: doesn't seem like this is used, why was it added as a dependency?

File: packages/vscode/package.json [hunk 3]
```diff
@@ -116,12 +125,15 @@
...
         "@salto-io/core": "0.2.0",
         "@salto-io/file": "0.2.0",
         "@salto-io/lang-server": "0.2.0",
         "@salto-io/workspace": "0.2.0",
+        "@types/openurl": "^1.0.0",
```
Comment (ori-moisis) [inline packages/vscode/package.json:133]: `@types` is only used for development, it should be a dev dependency, not a package dependency

File: packages/vscode/src/commands.ts [hunk 4]
```diff
@@ -42,3 +45,31 @@ export const createCopyReferenceCommand = (
...
   }
 }
+
+
+export const createGoToServiceCommand = (
+  workspace: ws.EditorWorkspace
+): (
+) => Promise<void> => async () => {
+  const editor = vscode.window.activeTextEditor
+  if (editor !== undefined) {
```
Comment (ori-moisis) [inline packages/vscode/src/commands.ts:55]: style nit - instead of writing all the code indented, when there is a quick "error flow", use it to return
in this case
```typescript
if (editor === undefined) {
  return
}
```
like the command above

File: packages/vscode/src/commands.ts [hunk 5]
```diff
@@ -42,3 +45,31 @@ export const createCopyReferenceCommand = (
...
+      workspace,
+      editor.document.fileName,
+      saltoPos
+    )
+
+    if (ctx.ref !== undefined) {
+      const url = await getElementUrl(workspace.workspace, ctx.ref.element.elemID)
+      if (url !== undefined) {
+        // Using this library instead of vscode.env.openExternal because of issue: https://github.com/microsoft/vscode/issues/112577
+        openurl.open(url.href)
```
Comment (ori-moisis) [inline packages/vscode/src/commands.ts:69]: in the `cli` package we seem to be using `https://www.npmjs.com/package/open` to open the browser, it seems a little more popular, any reason not to use it instead of `openurl`?
just seems like we are using it for exactly the same purpose, probably best not to have two different ways of doing the same thing

File: packages/vscode/src/commands.ts [hunk 6]
```diff
@@ -42,3 +45,31 @@ export const createCopyReferenceCommand = (
...
+    await workspace.awaitAllUpdates()
+    const saltoPos = vsPosToSaltoPos(position)
+    const ctx = await context.getPositionContext(
+      workspace,
+      editor.document.fileName,
+      saltoPos
+    )
+
+    if (ctx.ref !== undefined) {
+      const url = await getElementUrl(workspace.workspace, ctx.ref.element.elemID)
```
Comment (ori-moisis) [inline packages/vscode/src/commands.ts:66]: I have a bad feeling about this playing well with the whole `awaitAllUpdates` thing of the editor workspace.

@roironn - I think we will need your review on this and the way it is using the editor workspace, I think it is breaking some assumptions about what we can do with an editor workspace in async functions

File: packages/lang-server/src/workspace.ts (hunk 1 above)
Comment (noamhammer) [reply packages/lang-server/src/workspace.ts:23]: @ori-moisis You are right

File: packages/vscode/src/commands.ts (hunk 4 above)
Comment (alonstern) [reply packages/vscode/src/commands.ts:55]: The error flow here is to call `vscode.window.showErrorMessage('Go to service is not supported for the chosen token')`. I did it to avoid calling it multiple times

Comment (ori-moisis) [reply packages/vscode/src/commands.ts:55]: I honestly did not understand that flow until you wrote this. it is really hard to follow and quite unconventional that the success flow uses a `return` from the middle of the function.

calling that twice would be easier to follow IMO but another option to make the code structure more conventional is to separate the handling to two functions, one that should return the URL (that one can `return undefined` on each error scenario) and the other that calls it is decides what to do with the result (basically, if it got `undefined` it will show the error and if it got a url it will open it)

File: packages/lang-server/test/workspace.test.ts [hunk 7]
```diff
@@ -64,4 +64,9 @@ describe('workspace', () => {
     const removeNaclFilesMock = baseWs.removeNaclFiles as jest.Mock
     expect(removeNaclFilesMock.mock.calls[0][0]).toContain('all.nacl')
   })
+
+  it('should call workspace opearation', async () => {
```
Comment (ori-moisis) [inline packages/lang-server/test/workspace.test.ts:68]: this doesn't actually test anything, it just runs the code but doesn't validate that it does what it is supposed to.
for example, it doesn't make sure there can't be another operation that runs concurrently
it doesn't make sure the "operation" function actually gets the workspace

File: packages/vscode/src/events.ts [hunk 8]
```diff
@@ -57,6 +57,22 @@ export const onFileOpen = (): void => {
...
+export const onActiveTextEditorChange = async (
+  document: vscode.TextDocument | undefined,
+  workspace: ws.EditorWorkspace
+): Promise<void> => {
+  let isGoToServiceSupported: boolean
+  if (document === undefined || path.extname(document.fileName) !== FILE_EXTENSION) {
+    isGoToServiceSupported = false
+  } else {
+    const elements = await workspace.getElements(document.fileName)
+    isGoToServiceSupported = elements.some(e => e.elemID.adapter === 'salesforce')
```
Comment (ori-moisis) [inline packages/vscode/src/events.ts:69]: we should not have this hard coded here.
I don't think we need this feature in the vscode extension

File: packages/lang-server/src/workspace.ts [hunk 9]
```diff
@@ -184,6 +194,18 @@ export class EditorWorkspace {
...
-    if (this.runningSetOperation) await this.runningSetOperation
+    while (this.runningSetOperation !== undefined) {
+      // eslint-disable-next-line no-await-in-loop
+      await this.runningSetOperation
+    }
+  }
+
+
+  async runOperationWithWorkspace<T>(operation: WorkspaceOperation<T>): Promise<T> {
+    this.awaitAllUpdates()
```
Comment (ori-moisis) [inline packages/lang-server/src/workspace.ts:205]: discussed f2f why this won't work

File: packages/lang-server/src/workspace.ts [hunk 10]
```diff
@@ -184,6 +190,22 @@ export class EditorWorkspace {
...
+  }
+
+  async runOperationWithWorkspace<T>(operation: WorkspaceOperation<T>): Promise<T> {
+    while (this.runningWorkspaceOperation !== undefined) {
+      // eslint-disable-next-line no-await-in-loop
+      await this.awaitAllUpdates()
+    }
+    const operationPromise = operation(this.workspace)
+    this.runningWorkspaceOperation = this.waitForOperation(operationPromise)
+    await this.runningWorkspaceOperation
```
Comment (ori-moisis) [inline packages/lang-server/src/workspace.ts:208]: not sure we need this `await` here
doesn't hurt, but I don't think it helps

File: packages/lang-server/test/workspace.test.ts [hunk 11]
```diff
@@ -64,4 +65,35 @@ describe('workspace', () => {
...
+      arr.push('second')
+    }
+    const workspace = new EditorWorkspace(workspaceBaseDir, await mockWorkspace(naclFileName))
+    const firstPromise = workspace.runOperationWithWorkspace(firstOperation)
+    const secondPromise = workspace.runOperationWithWorkspace(secondOperation)
+
+    await firstPromise
+    await secondPromise
+
+    expect(_.isEqual(arr, ['second', 'second', 'first', 'first']) || _.isEqual(arr, ['first', 'first', 'second', 'second'])).toBeTruthy()
```
Comment (ori-moisis) [inline packages/lang-server/test/workspace.test.ts:97]: it should always be `first` before `second`

File: packages/lang-server/test/workspace.test.ts [hunk 12]
```diff
@@ -64,4 +65,35 @@ describe('workspace', () => {
     const removeNaclFilesMock = baseWs.removeNaclFiles as jest.Mock
     expect(removeNaclFilesMock.mock.calls[0][0]).toContain('all.nacl')
   })
+
+  it('should call workspace opearation', async () => {
+    const baseWs = await mockWorkspace(naclFileName)
+    const workspace = new EditorWorkspace(workspaceBaseDir, baseWs)
+    expect(await workspace.runOperationWithWorkspace(async _innerWorkspace =>
+      baseWs === _innerWorkspace)).toBeTruthy()
```
Comment (ori-moisis) [inline packages/lang-server/test/workspace.test.ts:73]: when testing a callback you should use a mock function.
that way you can check that the function was called with the expected parameters

you can see an example in [this snippet](https://gist.github.com/ori-moisis/e1c43261f4d2a2c5deeafde3f7b18bbc#file-jest_blog_2-ts-L25-L27) for the [blog post about jest](https://www.salto.io/post/typescript-unit-testing-pitfalls-with-jest-and-how-to-work-around-them)

**Review by ori-moisis — APPROVED** [review]
please fix the PR title (extra ":" in the ticket number)
